### ISSUE TITLE
1.5.8 not recognizing permissions and setting them to 0

### ISSUE THREAD

**Issue by JesseHC** [zenphoto/zenphoto#1301]
I noticed today after adding three videos with sidecar images that their thumbnails were not loading. When I checked the cache, they were there, that’s when I noticed that all thumbnails generated since updating from 1.5.7 to 1.5.8 were set with permissions to 0.

I was getting lots of the following errors in the logs-

```
NOTICE: Undefined index: CHMOD in /zenphoto/zp-core/functions-basic.php(127) : eval()'d code on line 82
eval called from require_once (functions-basic.php [127])
from require_once (admin-globals.php [12])
from admin-logs.php [8]
```

I tried to run setup again to see if it would fix it, but all of the setup files had also been set to 0, so setup failed.

I chmod the entire zenphoto directory to 755 for folders and 644 for files.

When I would run setup, it would say:
_File Permissions [are unknown (0)]_
![Zenphoto setup - permissions unknown 0](https://user-images.githubusercontent.com/82089671/123369183-2291e580-d532-11eb-86b6-c5dfd7400f08.png)

When I tried to apply the different file permissions in setup it would appear to timeout with a 500 internal server error message every time-
`https://example.com/zenphoto/zp-core/setup/index.php?xsrfToken=(long token)&chmod_permissions=1#`

Each time I tried it would set many of the folders to unusable permissions like 311 for the themes folder and basic folder under it. I would chmod everything again and try setup again.

I tried downloading 1.5.8 again and replacing the entire zp-core. Still the same issue.

What’s puzzling is that the permission value is still set to 644 in zenphoto.cfg.php in zp-data-

```
////////////////////////////////////////////////////////////////////////////////
if (!defined('CHMOD_VALUE')) { define('CHMOD_VALUE', $conf['CHMOD']); }


$conf['security_ack'] = 0644;
////////////////////////////////////////////////////////////////////////////////
```
This is on PHP 7.4.20.

**Comment by JesseHC** [comment]
I should add that each time I run setup I instantly get multiple debug logs with ```from setup::getResidentZPFiles (class-setup.php [39])``` spiraling through them all.

```
{840177:Thu, 24 Jun 2021 21:50:17 GMT} Zenphoto v1.5.8
  WARNING: opendir(/zenphoto/themes/basic): failed to open dir: Permission denied in /zenphoto/zp-core/setup/class-setup.php on line 33
  opendir called from setup::getResidentZPFiles (class-setup.php [33])
    from index.php [1285]
{840177:Thu, 24 Jun 2021 21:50:17 GMT}
  WARNING: readdir() expects parameter 1 to be resource, bool given in /zenphoto/zp-core/setup/class-setup.php on line 34
  readdir called from setup::getResidentZPFiles (class-setup.php [34])
    from index.php [1285]
{840177:Thu, 24 Jun 2021 21:50:17 GMT}
  WARNING: opendir(/zenphoto/themes/basic/): failed to open dir: Permission denied in /zenphoto/zp-core/setup/class-setup.php on line 33
  opendir called from setup::getResidentZPFiles (class-setup.php [33])
    from setup::getResidentZPFiles (class-setup.php [39])
      from index.php [1285]
{840177:Thu, 24 Jun 2021 21:50:17 GMT}
  WARNING: readdir() expects parameter 1 to be resource, bool given in /zenphoto/zp-core/setup/class-setup.php on line 34
  readdir called from setup::getResidentZPFiles (class-setup.php [34])
    from setup::getResidentZPFiles (class-setup.php [39])
      from index.php [1285]
{840177:Thu, 24 Jun 2021 21:50:17 GMT}
  WARNING: opendir(/zenphoto/themes/basic//): failed to open dir: Permission denied in /zenphoto/zp-core/setup/class-setup.php on line 33
  opendir called from setup::getResidentZPFiles (class-setup.php [33])
    from setup::getResidentZPFiles (class-setup.php [39])
      from setup::getResidentZPFiles (class-setup.php [39])
        from index.php [1285]
{840177:Thu, 24 Jun 2021 21:50:17 GMT}
  WARNING: readdir() expects parameter 1 to be resource, bool given in /zenphoto/zp-core/setup/class-setup.php on line 34
  readdir called from setup::getResidentZPFiles (class-setup.php [34])
    from setup::getResidentZPFiles (class-setup.php [39])
      from setup::getResidentZPFiles (class-setup.php [39])
        from index.php [1285]
{840177:Thu, 24 Jun 2021 21:50:17 GMT}
  WARNING: opendir(/zenphoto/themes/basic///): failed to open dir: Permission denied in /zenphoto/zp-core/setup/class-setup.php on line 33
  opendir called from setup::getResidentZPFiles (class-setup.php [33])
    from setup::getResidentZPFiles (class-setup.php [39])
      from setup::getResidentZPFiles (class-setup.php [39])
        from setup::getResidentZPFiles (class-setup.php [39])
          from index.php [1285]
{840177:Thu, 24 Jun 2021 21:50:17 GMT}
  WARNING: readdir() expects parameter 1 to be resource, bool given in /zenphoto/zp-core/setup/class-setup.php on line 34
  readdir called from setup::getResidentZPFiles (class-setup.php [34])
    from setup::getResidentZPFiles (class-setup.php [39])
      from setup::getResidentZPFiles (class-setup.php [39])
        from setup::getResidentZPFiles (class-setup.php [39])
          from index.php [1285]
{840177:Thu, 24 Jun 2021 21:50:17 GMT}
  WARNING: opendir(/zenphoto/themes/basic////): failed to open dir: Permission denied in /zenphoto/zp-core/setup/class-setup.php on line 33
  opendir called from setup::getResidentZPFiles (class-setup.php [33])
    from setup::getResidentZPFiles (class-setup.php [39])
      from setup::getResidentZPFiles (class-setup.php [39])
        from setup::getResidentZPFiles (class-setup.php [39])
          from setup::getResidentZPFiles (class-setup.php [39])
            from index.php [1285]
{840177:Thu, 24 Jun 2021 21:50:17 GMT}
  WARNING: readdir() expects parameter 1 to be resource, bool given in /zenphoto/zp-core/setup/class-setup.php on line 34
  readdir called from setup::getResidentZPFiles (class-setup.php [34])
    from setup::getResidentZPFiles (class-setup.php [39])
      from setup::getResidentZPFiles (class-setup.php [39])
        from setup::getResidentZPFiles (class-setup.php [39])
          from setup::getResidentZPFiles (class-setup.php [39])
            from index.php [1285]
```
![Multiple debug logs](https://user-images.githubusercontent.com/82089671/123372900-4dcc0300-d539-11eb-8909-cd4ba4e7328a.png)

This might be by design, but they are all around 5 MB.
![debug logs mostly around 5 MB](https://user-images.githubusercontent.com/82089671/123372946-64725a00-d539-11eb-809a-9859f10c46ac.png)

With forward slashes after /basic/ spiraling in error_log

```
[24-Jun-2021 14:50:17 America/Los_Angeles] PHP Warning:  opendir(/zenphoto/themes/basic): failed to open dir: Permission denied in /zenphoto/zp-core/setup/class-setup.php on line 33
[24-Jun-2021 14:50:17 America/Los_Angeles] PHP Warning:  readdir() expects parameter 1 to be resource, bool given in /zenphoto/zp-core/setup/class-setup.php on line 34
[24-Jun-2021 14:50:17 America/Los_Angeles] PHP Warning:  opendir(/zenphoto/themes/basic/): failed to open dir: Permission denied in /zenphoto/zp-core/setup/class-setup.php on line 33
[24-Jun-2021 14:50:17 America/Los_Angeles] PHP Warning:  readdir() expects parameter 1 to be resource, bool given in /zenphoto/zp-core/setup/class-setup.php on line 34
[24-Jun-2021 14:50:17 America/Los_Angeles] PHP Warning:  opendir(/zenphoto/themes/basic//): failed to open dir: Permission denied in /zenphoto/zp-core/setup/class-setup.php on line 33
[24-Jun-2021 14:50:17 America/Los_Angeles] PHP Warning:  readdir() expects parameter 1 to be resource, bool given in /zenphoto/zp-core/setup/class-setup.php on line 34
[24-Jun-2021 14:50:17 America/Los_Angeles] PHP Warning:  opendir(/zenphoto/themes/basic///): failed to open dir: Permission denied in /zenphoto/zp-core/setup/class-setup.php on line 33
[24-Jun-2021 14:50:17 America/Los_Angeles] PHP Warning:  readdir() expects parameter 1 to be resource, bool given in /zenphoto/zp-core/setup/class-setup.php on line 34
[24-Jun-2021 14:50:17 America/Los_Angeles] PHP Warning:  opendir(/zenphoto/themes/basic////): failed to open dir: Permission denied in /zenphoto/zp-core/setup/class-setup.php on line 33
[24-Jun-2021 14:50:17 America/Los_Angeles] PHP Warning:  readdir() expects parameter 1 to be resource, bool given in /zenphoto/zp-core/setup/class-setup.php on line 34
[24-Jun-2021 14:50:17 America/Los_Angeles] PHP Warning:  opendir(/zenphoto/themes/basic/////): failed to open dir: Permission denied in /zenphoto/zp-core/setup/class-setup.php on line 33
```

**Comment by acrylian** [comment]
The selector on  setup is a suggestion. You have to try which permissions work. What exactly works depends on your server. too strict ones cause trouble like you are encountering. Here is documented which the recommended permissions are: https://www.zenphoto.org/news/permissions-for-zenphoto-files-and-folders/

**Comment by JesseHC** [comment]
Yes, I’ve done that and read that.

The points I would like to make-
1. It was working up until 1.5.7.
2. It is not recognizing the zenphoto folder’s permissions and returning “unknown (0)”.
3. It is not recognizing the 0644 value set previously in zenphoto.cfg.php.
4. Trying to apply the permission changes crashes in setup every time, and in the process leaves multiple folders with weird unusable permissions like 311.

**Comment by acrylian** [comment]
I don't think and remember that we changed anything in this area at all from 1.5.7. I am sorry, I updated lots of installs on several hosts and had zero issues in this regards.

**Comment by acrylian** [comment]
The unknow persmissions results in the inability to read it because it is not allowed to do so. I can only recommend to remove all files and freshly upload them to perhaps fix this. Not overwrtiing, really manually deleting first.

**Comment by acrylian** [comment]
It can happen to mess up permissions via FTP. I remember to have had that years ago and I needed the host's help to fix this server site.

**Comment by JesseHC** [comment]
Well, that’s unfortunate. From past experience, if I have to rely on my host provider to troubleshoot or fix this, I’m probably out of luck.

I will say that I wasn’t having the permission issues on my MAMP test server before updating my live server.

**Comment by acrylian** [comment]
I am sorry, I probably have no other idea right now. Especially the permissions stuff in setup was really not changed… All I remember for changes in setup were some additional php extension requirement checks. 

So for now I can only recommend to do the fresh reupload of the core files to hopefully fix the issues by having the ftp upload correct them. That usually should help, unless it sets a wrong file owner. But if nothing changed on the server that should of course not happen either.

**Comment by JesseHC** [comment]
I tried uploading the files before using Cyberduck, having it change all of the permissions on upload to 755 for folders and 644 for files. I also tried uploading them and not changing the permissions, and the issue was still happening.

I also deleted the zp-data folder this time to start from scratch. When I did, I received an error message that it was unable to delete the generated tést.jpg. I was able to manually delete it in my host provider’s file manager.

Following your advice, I used a different FTP client, the one built-in to Dreamweaver this time, and the permissions issues went away.

Sorry for the hassle, I hadn’t experienced permissions issues like this before, and since it happened when I did the upgrade, I figured it might be related to that.

Marking this issue as closed, thanks for you time.
